src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,640 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.DocumentHighlighting;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Extensibility.NavigationBar;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.NavigateTo;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Protocol.LanguageServices.Extensions;
+using Microsoft.CodeAnalysis.QuickInfo;
+using Microsoft.CodeAnalysis.Structure;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using VSLocation = Microsoft.VisualStudio.LanguageServer.Protocol.Location;
+
+namespace Microsoft.CodeAnalysis.Protocol.LanguageServices
+{
+    /// <summary>
+    /// This is the roslyn language service.
+    ///
+    /// this glues workspace snapshot and language service protocol to
+    /// roslyn solution and services and at the end, it talks things in terms of
+    /// workspace snapshot id and LSP types.
+    /// </summary>
+    public sealed class RoslynLanguageService
+    {
+        // for now, it is null but later we might use this to provide info on types on dlls
+        private readonly IMetadataAsSourceFileService _metadataAsSourceService = null;
+
+        // TODO - Move hierarchicalDocumentSymbolSupport to client capabilities.
+        private readonly bool _hierarchicalDocumentSymbolSupport;
+        private readonly ClientCapabilities _clientCapabilities;
+
+        public RoslynLanguageService(ClientCapabilities clientCapabilities, bool hierarchicalDocumentSymbolSupport)
```

Comment:
will we contribute the hierarchicalDocumetSymbolSupport thing to LSP nuget as well? I think we should so that we can just use ClientCapabilites.

Reply:
Yep, working on that after I write some tests for this

Reply:
https://github.com/dotnet/roslyn/projects/45#card-20033973